### ISSUE TITLE
Error: write EPIPE and/or Error: read ECONNRESET running behind HAProxy 1.6

### ISSUE THREAD

**Issue by almccann** [mscdex/ssh2#483]
I am running a SSH server that spawns a child process to exec code (using `require('child_process').spawn`) after successful authentication.

The client server connections works fine on port 22 and connection is kept alive successfully through spawned process.

I am trying to setup up now with HAProxy 1.6, to forward port 22 to a non-privileged port on which the SSH server is listening.

However, after the child process is spawned the server either errors `Error: write EPIPE` or `Error: read ECONNRESET`.

Investigations into the HAProxy tcplog suggest HAProxy is performing normally.  Runing haproxy in debug mode I have:  
```
00000000:sshd.accept(0004)=0005 from [my ip]
00000000:ssh.srvcls[0005:0006]
00000000:ssh.clicls[0005:0006]
00000000:ssh.closed[0005:0006]
```
On the HAProxy tcplog  
`Oct 15 15:15:38 localhost haproxy[16036]: ip:51146 [15/Oct/2016:15:15:38.804] sshd ssh/ssh2server 1/0/+0 +0 -- 1/1/1/1/0 0/0`

The part of the server that is handling the streams:  
```
var channel = accept();
const git = spawn(info.command, { shell: true });

channel.stdin.pipe(git.stdin);
git.stdout.pipe(channel.stdout);
git.stderr.pipe(channel.stderr);
```

Any pointers or help would be awesome. Thanks in advance.

I'm using Node v6.1.0 and ssh2 v0.5.0.

**Comment by almccann** [comment]
Debug:  
```
Listening on port 7990
[418541.752565863] DEBUG: Local ident: 'SSH-2.0-ssh2js0.1.11srv'
[418541.752565863] DEBUG: Parser: IN_INIT (waiting for identification begin)
[418541.752565863] DEBUG: Parser: IN_INIT (waiting for rest of identification)
[418541.752565863] DEBUG: Parser: IN_GREETING
[418541.752565863] DEBUG: Parser: IN_HEADER
Client connected!
[418541.752565863] DEBUG: Remote ident: 'SSH-2.0-libssh2_1.6.0'
[418541.752565863] DEBUG: Outgoing: Writing KEXINIT
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: pktLen:644,padLen:7,remainLen:640
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXINIT
[418541.752565863] DEBUG: Comparing KEXINITs ...
[418541.752565863] DEBUG: (local) KEX algorithms: ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group14-sha1
[418541.752565863] DEBUG: (remote) KEX algorithms: diffie-hellman-group14-sha1,diffie-hellman-group-exchange-sha1,diffie-hellman-group1-sha1
[418541.752565863] DEBUG: KEX algorithm: diffie-hellman-group14-sha1
[418541.752565863] DEBUG: (local) Host key formats: ssh-rsa
[418541.752565863] DEBUG: (remote) Host key formats: ssh-rsa,ssh-dss
[418541.752565863] DEBUG: Host key format: ssh-rsa
[418541.752565863] DEBUG: (local) Client->Server ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com
[418541.752565863] DEBUG: (remote) Client->Server ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes256-cbc,rijndael-cbc@lysator.liu.se,aes192-cbc,aes128-cbc,blowfish-cbc,arcfour128,arcf
our,cast128-cbc,3des-cbc
[418541.752565863] DEBUG: Client->Server Cipher: aes128-ctr
[418541.752565863] DEBUG: (local) Server->Client ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com
[418541.752565863] DEBUG: (remote) Server->Client ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes256-cbc,rijndael-cbc@lysator.liu.se,aes192-cbc,aes128-cbc,blowfish-cbc,arcfour128,arcf
our,cast128-cbc,3des-cbc
[418541.752565863] DEBUG: Server->Client Cipher: aes128-ctr
[418541.752565863] DEBUG: (local) Client->Server HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1
[418541.752565863] DEBUG: (remote) Client->Server HMAC algorithms: hmac-sha1,hmac-sha1-96,hmac-md5,hmac-md5-96,hmac-ripemd160,hmac-ripemd160@openssh.com
[418541.752565863] DEBUG: Client->Server HMAC algorithm: hmac-sha1
[418541.752565863] DEBUG: (local) Server->Client HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1
[418541.752565863] DEBUG: (remote) Server->Client HMAC algorithms: hmac-sha1,hmac-sha1-96,hmac-md5,hmac-md5-96,hmac-ripemd160,hmac-ripemd160@openssh.com
[418541.752565863] DEBUG: Server->Client HMAC algorithm: hmac-sha1
[418541.752565863] DEBUG: (local) Client->Server compression algorithms: none,zlib@openssh.com,zlib
[418541.752565863] DEBUG: (remote) Client->Server compression algorithms: none
[418541.752565863] DEBUG: Client->Server compression algorithm: none
[418541.752565863] DEBUG: (local) Server->Client compression algorithms: none,zlib@openssh.com,zlib
[418541.752565863] DEBUG: (remote) Server->Client compression algorithms: none
[418541.752565863] DEBUG: Server->Client compression algorithm: none
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: pktLen:268,padLen:6,remainLen:264
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXDH_INIT
[418541.752565863] DEBUG: Outgoing: Writing KEXDH_REPLY
[418541.752565863] DEBUG: Outgoing: Writing NEWKEYS
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: pktLen:12,padLen:10,remainLen:8
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: NEWKEYS
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:28,padLen:10,remainLen:16
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: SERVICE_REQUEST
[418541.752565863] DEBUG: Outgoing: Writing SERVICE_ACCEPT (ssh-userauth)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:44,padLen:9,remainLen:32
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: USERAUTH_REQUEST (none)
[418541.752565863] DEBUG: Outgoing: Writing USERAUTH_FAILURE
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:348,padLen:13,remainLen:336
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: USERAUTH_REQUEST (publickey)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Outgoing: Writing USERAUTH_PK_OK
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:620,padLen:10,remainLen:608
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: USERAUTH_REQUEST (publickey)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Outgoing: Writing USERAUTH_SUCCESS
Client authenticated!
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:44,padLen:19,remainLen:32
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_OPEN (0, session)
[418541.752565863] DEBUG: Outgoing: Writing CHANNEL_OPEN_CONFIRMATION (r:0, l:0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:76,padLen:12,remainLen:64
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_REQUEST (0, exec)
Client wants to execute
[418541.752565863] DEBUG: Outgoing: Writing CHANNEL_SUCCESS (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Outgoing: Writing CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:156,padLen:9,remainLen:144
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:28,padLen:6,remainLen:16
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:28,padLen:16,remainLen:16
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:172,padLen:10,remainLen:160
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:28,padLen:16,remainLen:16
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:380,padLen:12,remainLen:368
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:28,padLen:16,remainLen:16
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:124,padLen:9,remainLen:112
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:28,padLen:16,remainLen:16
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:172,padLen:8,remainLen:160
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:28,padLen:17,remainLen:16
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:44,padLen:14,remainLen:32
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:44,padLen:19,remainLen:32
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
[418541.752565863] DEBUG: Parser: IN_PACKET
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: pktLen:28,padLen:15,remainLen:16
[418541.752565863] DEBUG: Parser: IN_PACKETDATA
[418541.752565863] DEBUG: Parser: Decrypting
[418541.752565863] DEBUG: Parser: HMAC size:20
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY
[418541.752565863] DEBUG: Parser: Verifying MAC
[418541.752565863] DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
[418541.752565863] DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (0)
[418541.752565863] DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: write EPIPE
    at exports._errnoException (util.js:949:11)
    at WriteWrap.afterWrite (net.js:793:14)
```

**Comment by mscdex** [comment]
What if you add an `'error'` event handler on the client object? Does that capture the error?

**Comment by almccann** [comment]
I am listening on the 'error' event handler on the client object- i.e. the listener in the ssh2 Server constructor.  I am chaining the event handlers though.  Will that bubble up from the session object?
    
```
.on('error', err => {
  console.log('An error occurred:', err);
})
```

I have also updated to v0.5.2 which you recommended in another EPIPE issue.  What version of Node.js do you recommend?

**Comment by mscdex** [comment]
I'm not sure I follow, can you post more context where you're adding that error handler?

As far as recommending node.js versions, the newer the better of course as a general rule (unless you need LTS), but it should work with all node versions (v0.10 and up).

**Comment by mscdex** [comment]
If the server works fine outside of haproxy and has issues when proxying via haproxy, then I'm not sure what to suggest as I am not familiar enough with haproxy (other than make sure you attach error handlers).

**Comment by almccann** [comment]
This is the main part:  
```
new ssh2.Server({
	hostKeys: [fs.readFileSync( config.get( 'hostKeyPath' ) )]//,
	// debug: (cfg) => { console.log(cfg); }
}, (client) => {
	console.log('Client connected!');
	client.on('authentication', (ctx) => {
		// Connection logic
	})
	.on('ready', () => {
		console.log('Client authenticated!');
	})
	.on('session', (accept, reject) => {
		var session = accept();
		session.on('exec', (accept, reject, info) => {
			console.log('Client wants to execute: ' + inspect(info.command));
                        // Session logic
                        // spawned process here
		})
		.on('close', () => {
			console.log('The session was closed');
		});
	})
	.on('error', (err) => {
		console.log('An error occurred:', err);
	})
	.on('end', () => {
		console.log('The client socket disconnected');
	})
	.on('close', (hadError) => {
		console.log(`The client socket was closed, with an error? ${hadError}`);
	});
})
.listen(config.get( 'port' ), '127.0.0.1', function () {
	console.log('Listening on port ' + this.address().port);
});
```

I'm trying now to forward from port 22 with NAT tables now, and still having the unhandled error messages.

**Comment by mscdex** [comment]
That portion looks fine to me for the most part. Do you have some way for me to easily replicate your test environment?

**Comment by almccann** [comment]
Or are you interested in the OS?

**Comment by mscdex** [comment]
I was thinking more like a set of instructions or configuration, like the NAT entries or haproxy config, what OS and OS version, what node version, etc., or if you have a docker container or something similar that can be used to replicate the error.

**Comment by almccann** [comment]
Okay.  Unfortunately I'm not using containers with this at the moment to share, so I'll provide you with as much information on the environment and configurations as possible.  Apologies if any of this info is self-explanatory for you.

```
Distributor ID: Debian
Description:    Debian GNU/Linux 8.3 (jessie)
Release:        8.3
Codename:       jessie
```
`Linux version 3.16.0-4-amd64 (debian-kernel@lists.debian.org) (gcc version 4.8.4 (Debian 4.8.4-1) ) #1 SMP Debian 3.16.7-ckt20-1+deb8u3 (2016-01-17)`

**The HAProxy configuration:**  
Here the ssh2 server is aliased `backend ssh` which I am running on port 7990.
```
global
  daemon
  maxconn 10000
  log 127.0.0.1 local0

defaults
  log global
  option tcplog
  option logasap
  timeout connect 500s
  timeout client 5000s
  timeout server 2h
  timeout server-fin 5000s
  timeout client-fin 5000s
  timeout tunnel 5000s
  option tcpka

frontend sshd
  bind *:22
  default_backend ssh
  timeout client 2h

backend ssh
  mode tcp
  server ssh2server 127.0.0.1:7990 check port 7990
```

**The NAT configuration.**
I am using `ufw` to create the rules, although the configuration can be enabled with `iptables` and I believe would be identical.

The NIC of my system is `eth0`.
`ip -4 addr show scope global` will list the interfaces.
`ip route show | grep default` will confirm the public.

Enable packet forwarding:
Edit `/etc/default/ufw`: change forward policy to `DEFAULT_FORWARD_POLICY="ACCEPT"`
Edit `/etc/ufw/sysctl.conf` and uncomment: `net/ipv4/ip_forward=1`
Enable routing to localhost:  `sysctl -w net.ipv4.conf.eth0.route_localnet=1`
Add before rules in `/etc/ufw/before.rules` at the head of the file:
```
*nat
:PREROUTING ACCEPT [0:0]
-A PREROUTING -i eth0 -p tcp --dport 22 -j DNAT --to-destination 127.0.0.1:7990
COMMIT
```
Enable rules: `sudo ufw disable && sudo ufw enable`.

There is a bug in `ufw` that may result in the rules not saving.  To overcome this reboot the system.  To check: `sudo iptables -L -vt nat`.

I have tried with node versions v6.9.0 and v6.1.0. Errors presented with both.

**Comment by mscdex** [comment]
What version of HAProxy did you test with?

**Comment by mscdex** [comment]
Ok I have a setup with debian jessie, haproxy 1.6.9, node v6.9.1, and the code you've provided, but I haven't been able to make it crash with that error yet. Is it perhaps a particular command you're running (git)? Do other commands work fine?

**Comment by almccann** [comment]
Yes, I am using HAProxy 1.6.9.

In the server session, I have:  
```
.on('session', (accept, reject) => {
		var session = accept();
		session.on('exec', (accept, reject, info) => {
			console.log('Client wants to execute: ' + inspect(info.command));
			// returns a streams2 Duplex Stream (channel) for the command execution
			// channel.stderr is a writable stream

			// ensure command is a git command
			if (info.command.indexOf('git') === 0) {
				var channel = accept();
				const git = spawn(info.command, { shell: true });

				channel.stdin.pipe(git.stdin);
				git.stdout.pipe(channel.stdout);
				git.stderr.pipe(channel.stderr);

				git.on('close', function(code, signal) {
				  channel.close();
				  console.log(`git exited with code ${code} and signal ${signal}`);
				})
				.on('error', (err) => {
					console.log('An error occurred:', err);
				})
				.on('exit', function(code, signal) {
				  channel.exit(code, signal);
				});
			} else {
				channel.close();
				console.log(`info command did not contain git`);
			}
		})
```
(I know this is insecure, but its for a proof-of-concept).

If you add the server as a remote in a local folder (`ssh://user-who-runs-server@yourserver.com/path/to/bare.git`), and init the bare repo at the path on the server you specify.

Then, I am attempting to `git push` from the local repository to port 22 with this server running ssh2 on non-privileged port.

**Comment by mscdex** [comment]
I just tried that and I can push to the remote repo without any problems :-\

**Comment by mscdex** [comment]
I've also tried using the iptables method, but for whatever reason wasn't able to get it to connect properly. I don't know if this has something to do with the fact that I'm using this under a local VM or not.

FWIW is running the node script as root, binding to the low level port (22), and then passing a normal/non-root user name/id to `process.setgid()` and `process.setuid()` to drop privileges a possible alternative solution? What about using authbind? In addition to haproxy, both of these solutions also worked for me when I tested them.
